src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/PackageDeclarationCheck.java

```diff
@@ -72,12 +93,30 @@ public void finishTree(DetailAST ast) {
             if (ast != null) {
                 lineNumber = ast.getLineNo();
             }
-            log(lineNumber, MSG_KEY);
+            log(lineNumber, MSG_KEY_MISSING);
         }
     }
 
     @Override
     public void visitToken(DetailAST ast) {
         defined = true;
+
+        if (!ignoreMismatch) {
+            // Calculate the directory name, but stripping off the last
+            // part.
+            final String fileName = getFileContents().getFileName();
+            final int lastPos = fileName.lastIndexOf(File.separatorChar);
+            final String directoryName = fileName.substring(0, lastPos);
+
+            // Get the package name and format it like a directory name.
```

Comment:
```java
final DetailAST packageNameAst = ast.getLastChild().getPreviousSibling();
final FullIdent fullIdent = FullIdent.createFullIdent(packageNameAst);
final String packageName = fullIdent.getText().replace('.', File.separatorChar);
```
I cannot create a method `getPackageName(ast)` as I also need `fullIdent.getLineNo() ` .If this is not acceptable to you I'll create a global variable of `FulllIdent` and update it's object?